### PR TITLE
Multiple graphs one page

### PR DESCRIPTION
**Pull request by jez500**
Hi @tobybellwood & @jorgegc 

A few more improvements and fixes
* Bugfix: Fixed issue when more than one DVF exists on same edit form
* Updated readme (as requested by sponsors)
* Added ability to set a default colour palette on chart visualisation
* Added ability for 3rd party modules to override settings and documented via api

Cheers!

### CODE REVIEW

File: README.md
Comment (tobybellwood) [inline README.md]: Technically if we're changing it, it should be "Department of the Environment and Energy" (and https the link)

**Comment by tobybellwood** [comment]
Thanks @jez500 - I'll wait for @jorgegc to code review if poss?
It's looking good!

File: src/Plugin/Field/FieldFormatter/VisualisationUrlFormatter.php
Comment (jorgegc) [inline src/Plugin/Field/FieldFormatter/VisualisationUrlFormatter.php]: Line break before return improves readability

File: src/Plugin/Field/FieldFormatter/VisualisationUrlFormatter.php
Comment (jorgegc) [inline src/Plugin/Field/FieldFormatter/VisualisationUrlFormatter.php]: Same as line 50.

File: src/Plugin/Field/FieldFormatter/VisualisationUrlFormatter.php
Comment (jorgegc) [inline src/Plugin/Field/FieldFormatter/VisualisationUrlFormatter.php]: Same as line 50.

File: src/FieldWidgetTrait.php
Comment (jorgegc) [inline src/FieldWidgetTrait.php]: Line break before return improves readability

File: src/Plugin/Field/FieldFormatter/VisualisationFileFormatter.php
Comment (jorgegc) [inline src/Plugin/Field/FieldFormatter/VisualisationFileFormatter.php]: How do you feel about extending both this class and `VisualisationUrlFormatter` from a common class `VisualisationChartFormatter`.

Alternatively, how do you feel about creating a single field formatter that both the URL and File field types use as their default formatters (e.g. `VisualisationChartFormatter`)? I am leaning more towards this option...

File: src/Plugin/Field/FieldFormatter/VisualisationUrlFormatter.php
Comment (jorgegc) [inline src/Plugin/Field/FieldFormatter/VisualisationUrlFormatter.php]: Lines 27-30 are too indented. This has got me a number of times too!

File: src/Plugin/Field/FieldWidget/VisualisationFileWidget.php
Comment (jorgegc) [inline src/Plugin/Field/FieldWidget/VisualisationFileWidget.php:179]: I am not sure this changes anything in your solution but please remember that this needs to be optional. Hopefully, in the not-too-distant future we will be able to create visualisations with data coming from Views.

File: src/Plugin/Visualisation/Style/VisualisationStyleBase.php
Comment (jorgegc) [inline src/Plugin/Visualisation/Style/VisualisationStyleBase.php]: Can we please inject `module_handler` via the service container?

File: src/Plugin/Visualisation/Style/VisualisationStyleBase.php
Comment (jorgegc) [inline src/Plugin/Visualisation/Style/VisualisationStyleBase.php]: Same as line 48.

File: src/Plugin/Visualisation/Style/VisualisationStyleBase.php
Comment (jorgegc) [inline src/Plugin/Visualisation/Style/VisualisationStyleBase.php]: Same as line 48.

File: src/Plugin/Visualisation/VisualisationBase.php
Comment (jorgegc) [inline src/Plugin/Visualisation/VisualisationBase.php]: As I mentioned in one of my previous comments, this should be an optional thing if we want DVF to work with non-entity data sources.

**Review by jorgegc — CHANGES_REQUESTED** [review]
Hey @jez500 PR looks good but I've picked up on a couple of things that we should probably discuss before merging.

cc @tobybellwood

File: src/Plugin/Field/FieldWidget/VisualisationFileWidget.php
Comment (jez500) [reply src/Plugin/Field/FieldWidget/VisualisationFileWidget.php:179]: Hey @jorgegc I added this to provide more value to the `hook_dvf_xxxx_alter` it has no immediate use but provides a huge amount of flexibility if themes/modules want to alter based on other field values, especially handy when used with govCMS. 

A lot more than this would need refactoring if it was optional as this method relies on `FieldItemListInterface` which is specific to an entity field item.

That said, I agree it should be optional in the vis plugin which i'll ensure will have a graceful fallback.

File: src/Plugin/Visualisation/Style/VisualisationStyleBase.php
Comment (jez500) [reply src/Plugin/Visualisation/Style/VisualisationStyleBase.php]: I thought about that but was proving a bit painful to use Dep injection for this (and I am low on time) also
* [following suite](https://github.com/jez500/dvf/blob/3395d59c0c84ff1dd1b13895fddb05d1e7ed6de6/src/Plugin/Field/FieldWidget/VisualisationUrlWidget.php#L98) :P 
* It's very common to see this used without dep injection in [many places in core](https://github.com/drupal/drupal/blob/ba42166da282f6c727d15869d0b3702dd309fc91/core/modules/node/src/NodeViewBuilder.php#L106)

File: README.md
Comment (jez500) [reply README.md]: Thanks Toby, good catch and updated now 👍

Comment (tobybellwood) [reply README.md]: hey - I can't call out much on the code side - so I'll play to my strengths!

File: src/Plugin/Field/FieldFormatter/VisualisationFileFormatter.php
Comment (jez500) [reply src/Plugin/Field/FieldFormatter/VisualisationFileFormatter.php]: Good idea! I refactored to be `VisualisationDefaultFormatter` and removed file/url formatters. Feels better having a default that new formatters can extend.

File: src/Plugin/Field/FieldWidget/VisualisationFileWidget.php
Comment (jez500) [reply src/Plugin/Field/FieldWidget/VisualisationFileWidget.php:179]: Updated `VisualisationStyleInterface` to indicate this could be null. Have left it in this file for above reasons :)

File: src/Plugin/Visualisation/Style/VisualisationStyleBase.php
Comment (jez500) [reply src/Plugin/Visualisation/Style/VisualisationStyleBase.php]: I caved... updated

**Comment by jez500** [comment]
Cheers for the feedback @jorgegc items actioned + a few more cs fixes

**Review by jorgegc — APPROVED** [review]

**Comment by jorgegc** [comment]
Nice work @jez500, thanks!

**Comment by tobybellwood** [comment]
nice one @jez500 

should we start planning to get an alpha/beta release out - I'll create an issue?

**Pull request merged**
